### DISCUSSION TITLE
Starting point - Rocktron All Access

### DISCUSSION THREAD

**Discussion by DavidOpgh** [RomanKubiak/ctrlr#457] [General]
I want to create a panel for the Rocktron All Access midi footpedal. 

It has minimal sysex implementation and little available documentation. 
It only supports Sysex dumps and uploads of the entire memory so I'd like to dump the sysex, create/modify patches, and upload the modified sysex.

Is there an existing panel I can use as a starting point? 
I haven't done a deep dive but it seems like a lot of panels dump/edit/load individual patches, not an entire sysex memory dump.

Is there documentation / code examples that show how to parse the sysex dump into patches?

Are there code examples that show how to modify the speed of sysex uploads? 
The Rocktron is very slow handling uploads. It requires a delay after sending a block of data or you'll get a buffer overflow. Using the Midi-OX app I've been able to configure the parameters to upload sysex files reliably. I want to duplicate that in Ctrlr

thanks!

**Comment by dnaldoog** [comment]
Hi @DavidOpgh 

Can you upload an example dump file as sysex?

**Reply by DavidOpgh** [reply]
Hello @dnaldoog ! 
Thanks for the quick reply.
Here are two sysex dumps and the users manual which is the only documentation available. Unfortunately the manual doesn't contain the data structure of the sysex. I've just started to reverse engineer the structure.

The manual does provide some clues what to expect in the sysex data.
120 presets, each of which can perform
- up to 16 program change commands
- up to 15 control changes
- up to 5 other MIDI commands
- system exclusive message up to 30 bytes in length.
- preset titles can be up to 13 characters in length.

The sysex dump begins with 120 blocks of data each 309 bytes long. 
I haven't determined what data follows the presets, but it's probably the Midi, Song, and Setup data mentioned in the manual. 

[allaccess.pdf](https://github.com/RomanKubiak/ctrlr/files/8521199/allaccess.pdf)
[All Access - default.zip](https://github.com/RomanKubiak/ctrlr/files/8521213/All.Access.-.default.zip)
[All Access - 04-13-2022.zip](https://github.com/RomanKubiak/ctrlr/files/8521218/All.Access.-.04-13-2022.zip)

**Comment by dnaldoog** [comment]
@DavidOpgh 

Thanks for that - I was able to find where the patch names are in the __All Access - 04-13-2022.zip__ file. They start at the 8th byte but there seems to be an empty 00 byte between each ASCII binary number, so for a 12 letter name, which seems to be the limit, there are 24 bytes of data. Each patch has its own message so you should be able to slow down the transmission of each message in Ctrlr, but I'm not sure how to slow the individual byte rate transfer in Ctrlr. 

What each byte of data represents would be a case of trial and error, finding what each byte does, but this should give you a start. But, it looks like you can use a similar method to extract the data as with the name.

<pre>
--
-- Called when a panel receives a midi message (does not need to match any modulator mask)
-- @midi   CtrlrMidiMessage object
--
sf = string.format
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local s = midi:getSize()

    panel:getLabelComponent("d"):setText(sf("received %d bytes", s))

    if s == 309 then
        extractName(midi, s)
    end
end

extractName = function(midi, size)
    local nameData, tmp = MemoryBlock(), MemoryBlock() -- there's a zero byte between  each Ascii character
    tmp:loadFromHexString(midi:getData():getRange(8, 24):toHexString(1))
    local t = {}
    for i = 0, tmp:getSize() - 1, 2 do -- only access every second byte
        table.insert(t, tmp:getByte(i))
    end
    nameData:createFromTable(t)

    panel:getLabelComponent("d"):append(sf("\n%s", nameData:toString()))
end
</pre>

[Rocktron All Access_1.zip](https://github.com/RomanKubiak/ctrlr/files/8522009/Rocktron.All.Access_1.zip)

**Comment by dnaldoog** [comment]
Version 2 has a timer and sends out the 309 byte message according to the ms time you set on the panel.

[Rocktron All Access_2.zip](https://github.com/RomanKubiak/ctrlr/files/8522273/Rocktron.All.Access_2.zip)

![rocktronSS](https://user-images.githubusercontent.com/76831683/164252213-c67c2c0b-5a48-46bb-834f-ff0dbcdc83ac.png)

**Comment by DavidOpgh** [comment]
@dnaldoog 
Thanks for the panels. I gave them a quick look but I'll review them more a little later. 

I'm having issues today with reliable sysex uploads using MIDI-OX. I currently have it configured with an output buffer of 32 bytes and waiting 200ms between each buffer but was still getting a buffer overflow error. If the Rocktron can't handle an upload of the entire patch at full speed I'll probably have to add some additional code to break up the sysex into smaller blocks. 

Looking at the display on the Rocktron I believe patch names start at byte x0006 (7th byte), The names normally begin with a space x20 after the patch number. Looking at the sysex each character has a trailing byte of x00 which gives a total of 13 characters for names.

Yes, it looks like it's going to be a lot of trial and error trying to figure out what each byte does. I believe a lot of the data is similar so once I figure out the data for one function does it should apply to multiple locations.

**Reply by dnaldoog** [reply]
> I'll probably have to add some additional code to break up the sysex into smaller blocks.

I'm not sure if this can be done in Ctrlr.

- https://ctrlr.org/forums/topic/splitting-sysex-for-older-synths/
- https://ctrlr.org/forums/topic/splitting-sysex-for-older-synths/page/2/#post-115272

**Reply by DavidOpgh** [reply]
@dnaldoog 
Thanks for the info. This does appear similar to my issue. 
The Rocktron doesn't appear to handle processing an entire sysex block. I'm going to do more testing to confirm.

If Ctrlr can't be coded to send smaller blocks of data it would acceptable if could get Ctrlr to capture the dump and edit/create the sysex file. I could then use something like MIDI-OX to upload the sysex file.

**Reply by dnaldoog** [reply]
That’s right. I haven’t tried it but I have thought one could send Ctlr midi out to midi-ox or Bome SendSX then engage THRU thereby throttling the speed and then sending to the device.

**Reply by DavidOpgh** [reply]
@dnaldoog 
Yes, I could see how that might work.
I haven't tried using Bome Send SX yet. I'll have to give it a look.

**Reply by dnaldoog** [reply]
I prefer it - the interface is much less confusing and cluttered than Midi-Ox 😊

**Reply by DavidOpgh** [reply]
I tried Bome Send SX and it wasn't able to slow the transfer rate enough for the All Access. The slowest Send SX runs is 1/10 speed and I'm guessing I need the rate to be about 1/50 speed.  However I was able to reconfigure MIDI-OX to consistently upload sysex files by sending it in blocks of 16 bytes with a delay of 200ms in between blocks.

**Reply by dnaldoog** [reply]
> I tried Bome Send SX and it wasn't able to slow the transfer rate enough for the All Access. The slowest Send SX runs is 1/10 speed and I'm guessing I need the rate to be about 1/50 speed. However I was able to reconfigure MIDI-OX to consistently upload sysex files by sending it in blocks of 16 bytes with a delay of 200ms in between blocks.

I tried that too and yes _Bome Send Sx_ doesn't slow down enough, whereas _Midi-OX_ does! I still wonder if this can't be done in Ctrlr, because the author seemed to think of everything else?
